src/core/service.c

```diff
@@ -1440,6 +1440,93 @@ static bool service_exec_needs_notify_socket(Service *s, ExecFlags flags) {
         return s->notify_access != NOTIFY_NONE;
 }
 
+static int service_create_monitor_md_env(Job *j, char **ret) {
+        _cleanup_free_ char *var = NULL;
+        const char *list_delim = ";";
+        bool first = true;
+        Unit *tu;
+
+        assert(j);
+        assert(ret);
+
+        /* Create an environment variable 'MONITOR_METADATA', if creation is successful
+         * a pointer to it is returned via ret.
+         *
+         * This variable contains a space separated set of fields which relate to
+         * the service(s) which triggered job 'j'. Job 'j' is the JOB_START job for
+         * an OnFailure= or OnSuccess= dependency. Format of the MONITOR_METADATA
+         * variable is as follows:
+         *
+         * MONITOR_METADATA="SERVICE_RESULT=<result-string0>,EXIT_CODE=<exit-code0>,EXIT_STATUS=<exit-status0>,
+         *                   INVOCATION_ID=<id>,UNIT=<triggering-unit0.service>;
+         *                   SERVICE_RESULT=<result-stringN>,EXIT_CODE=<exit-codeN>,EXIT_STATUS=<exit-statusN>,
+         *                   INVOCATION_ID=<id>,UNIT=<triggering-unitN.service>"
+         *
+         * Multiple results may be passed as in the above example if jobs are merged, i.e.
+         * some services a and b contain an OnFailure= or OnSuccess= dependency on the same
+         * service.
+         *
+         * For example:
+         *
+         * MONITOR_METADATA="SERVICE_RESULT=exit-code,EXIT_CODE=exited,EXIT_STATUS=1,INVOCATION_ID=02dd868af2f344b18edaf74b618b2f90,UNIT=failer.service;
+         *                   SERVICE_RESULT=exit-code,EXIT_CODE=exited,EXIT_STATUS=1,INVOCATION_ID=80cb228bd7344f77a090eda603a3cfe2,UNIT=failer2.service"
+         */
+
+        LIST_FOREACH(triggered_by, tu, j->triggered_by) {
+                Service *env_source = SERVICE(tu);
+                int r;
+
+                if (!env_source)
+                        continue;
+
+                if (first) {
+                        /* Add the environment variable name first. */
+                        r = strextendf(&var, "MONITOR_METADATA=");
```

Comment:
don't use format strings if all you want is add strings. formatting via format strings is *slow*...

just use plain strextend